pkg/common/exec.go

```diff
@@ -0,0 +1,93 @@
+package common
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+)
+
+// StreamingExec models a generic command execution that consumers can use to
+// execute commands and stream their output to an io.Writer. For example
+// compute commands can use this to standardize the flow control for each
+// compiler toolchain.
+type StreamingExec struct {
+	command string
+	args    []string
+	env     []string
+	verbose bool
+	output  io.Writer
+}
+
+// NewStreamingExec constructs a new StreamingExec instance.
+func NewStreamingExec(cmd string, args, env []string, verbose bool, out io.Writer) *StreamingExec {
+	return &StreamingExec{
+		cmd,
+		args,
+		env,
+		verbose,
+		out,
+	}
+}
+
+// Exec executes the compiler command and pipes the child process stdout and
+// stderr output to the supplied io.Writer, it waits for the command to exit
+// cleanly or returns an error.
+func (s StreamingExec) Exec() error {
+	// Construct the command with given arguments and environment.
+	//
+	// gosec flagged this:
+	// G204 (CWE-78): Subprocess launched with variable
+	// Disabling as the variables come from trusted sources.
+	/* #nosec */
+	cmd := exec.Command(s.command, s.args...)
+	cmd.Env = append(os.Environ(), s.env...)
+
+	// Pipe the child process stdout and stderr to our own output writer.
```

Comment:
As I understand it, the requirements here are

1. Child process stdout and stderr should stream to s.output in real time
2. In case of error and --verbose, stderr should be captured and printed separately

Is that true? Assuming so, I think we might be able to avoid the pipes altogether..

```
cmd := exec.Command(s.command, s.args...)
cmd.Env = append(os.Environ(), s.env...)

var stderrBuf bytes.Buffer
cmd.Stdout = s.output
cmd.Stderr = io.MultiWriter(s.output, stderrBuf)

if err := cmd.Run(); err != nil {
    if s.verbose && stderrBuf.Len() > 0 {
        return ...
    }
    return ...
}

return nil
```

WDYT?

Reply:
Nice. Have implemented and tested in: https://github.com/fastly/cli/pull/160/commits/b6776d3b6105ed6df7cc9e03c2d421392933d602